### PR TITLE
Withhold the automatic Metal 4 tensor enable on M5 until accumulate parity

### LINKED ISSUE #946: M5 Metal 4 TensorOps prefill drifts from reference kernels by whole logits
## Summary

On M5-class GPUs the automatic Metal 4 TensorOps route (`tensor_matmul=on`) produces prompt logits that drift from the simdgroup reference kernels by whole logit units — far beyond reduction-order noise. Greedy streams diverge within a couple of tokens at ≥32-token prefills, `./ds4_test --metal-tensor-equivalence` fails on M5 hardware, and M5 vs pre-M5 machines give different eval output for identical binaries and model bytes.

## Repro (M5 Max, macOS 27.0, current main, GLM-5.3-Flash-Q2)

```sh
# the repo's own parity gate
DS4_TEST_MODEL=gguf/GLM-5.3-Flash-Q2.gguf \
DS4_TEST_VECTOR_FILE=tests/test-vectors/glm-openrouter/official.vec \
  ./ds4_test --metal-tensor-equivalence
# -> metal-tensor-equivalence: ERR, 28 failures
#    worst_rms=1.38592 worst_max_abs=7.26952 top1_mismatch=2

# cross-machine greedy dump (run on M5 and on any pre-M5 machine, then diff)
./ds4 -m gguf/GLM-5.3-Flash-Q2.gguf --metal --nothink -sys "" --temp 0 \
  -n 2 --ctx 32768 --prompt-file <prompt ≥ 32 tokens> \
  --dump-logprobs /tmp/dump.json --logprobs-top-k 20
```

## Measurements

Control: same git commit on both machines, byte-identical model file (size + head/tail hashes), same macOS 27.0; Metal shaders compile at runtime from identical source. Both machines are individually deterministic (byte-identical reruns).

- M5 with `DS4_METAL_DISABLE_METAL4=1` is **bit-identical** to M3 Ultra: greedy logprob dumps, `--decode-consistency` numbers (0.299938202/0.0643299961 — same digits), and the full 2400-token `ds4-eval` case-6 trace.
- M5 default (tensor route): 58-token prompts already drift 0.8–1.7 logits over the common top-k and flip greedy argmax; a 3778-token prompt flips the first sampled token ("The" → "the") and all 16 continuation steps diverge.
- `--decode-consistency` at a 3794-token prefix: M5 max_abs 1.004 / rms 0.203 vs 0.300 / 0.064 with the tensor route disabled.
- Bisected by disabling one MPP dispatch site at a time (dense NAX matmul, routed-MoE MPP swap, indexer NAX scores, attention-out MPP): the error sits in the `matmul2d` tiles themselves — both the staged routed-MoE tiles (`kernel_mul_mm_id_*_mpp`) and the direct-RHS dense prefill tiles contribute; the indexer score path measured bit-exact.
- The error is invariant to operand staging precision (half or float staging both drift identically in scale) and to `matmul2d_descriptor::relaxed_precision` (both values produce identical output). That isolates it to the driver/hardware TensorOps accumulate for these shapes, which we cannot correct from MSL.

## Impact

- Cross-machine eval and quality comparisons between M5 and pre-M5 machines are confounded (e.g. `ds4-eval` case 6/aime2025-16 picks 28 on M3 Ultra vs 24 on M5 Max with the tensor route).
- The kernels are model-family shared, so the DeepSeek IQ2_XXS/Q2_K routed prefill is exposed on M5 as well.

## Proposed fix (PR follows)

Withhold the automatic tensor enable on M5-class devices until the accumulate matches the reference kernels; keep `DS4_METAL_ENABLE_TENSOR=1` as the explicit opt-in for performance work. With the withhold, the M5 becomes bit-identical to the M3 Ultra across all measured fixtures and the tensor-equivalence gate passes. A/B on the 3778-token prompt showed no measurable prefill wall-time cost on GLM-5.3-Flash-Q2 (legacy 16.0–17.2 s vs tensor 15.3–17.7 s, warm cache). The PR also keeps `--metal-tensor-equivalence` meaningful: the `auto` route asserts parity, and an opt-in `tensor-optin` reporting run keeps the accumulate gap visible in QA logs.

## Side notes (GLM drift tooling)

- `--metal-graph-test` rejects GLM-5.3 ("unexpected layer-0 tensor layout" — it expects DSA attention at layer 0, but GLM-5.3 layer 0 is KDA linear attention).
- `--cpu --first-token-test` segfaults (exit 139) for GLM-5.3.
- The GLM OpenRouter official vectors target `z-ai/glm-5.2`, so no official 5.3 reference vectors exist yet.

### PR DESCRIPTION
**Pull request by ivanfioravanti**
Closes #946.

## What

Stops auto-enabling the Metal 4 TensorOps route (`tensor_matmul`) on M5-class devices, keeping `DS4_METAL_ENABLE_TENSOR=1` as an explicit opt-in. The automatic route was numerically lossy on M5 hardware (issue #946): the `matmul2d` accumulate drifts from the simdgroup reference kernels by whole logit units, independent of operand staging precision and of `matmul2d_descriptor::relaxed_precision`.

## Verification (M5 Max + M3 Ultra, macOS 27.0, GLM-5.3-Flash-Q2, same commit + byte-identical model)

| Check | Before (tensor auto) | After (withheld) |
|---|---|---|
| `ds4_test --metal-tensor-equivalence` | ERR, 28 failures, worst_rms=1.39, worst_max_abs=7.27 | **OK** (auto route all-zero deltas) |
| Greedy logprob dumps vs M3 Ultra (58 / 309 / 3778-token prompts) | 0.8–3.5+ logit drift, argmax flips | **byte-identical** |
| `ds4-eval` case 6 (aime2025-16, 2400 greedy tokens) | M5 picks 24, M3 picks 28, outputs diverge at token 2 | **byte-identical traces** (both pick 28) |
| `--decode-consistency` (3794-token prefix) | max 1.004 / rms 0.203 | 0.300 / 0.064 — same digits as M3 |
| 3778-token prefill wall time (A/B, warm cache) | 15.3–17.7 s | 16.0–17.2 s (no measurable cost) |

`DS4_METAL_ENABLE_TENSOR=1` still re-enables the route and reproduces the old tensor output bit-for-bit (verified).

## Gate change

With the route off by default, `--metal-tensor-equivalence` would have become legacy-vs-legacy and vacuous. The test now:

- asserts parity for the shipped `auto` route (must stay all-zero on every GPU), and
- runs an explicit `tensor-optin` candidate in **reporting mode** so the accumulate gap stays visible in QA logs (`worst_rms=1.386 worst_max_abs=7.27` on M5 today) without gating the release.

Also adds the gate to `QA_BEFORE_RELEASES.md` with a note that M5 prefill throughput references predate the withhold.

## Notes for reviewers

- The routed-MoE MPP kernels, dense NAX kernels, and `DS4_METAL_HAS_TENSOR` shader variants are all unchanged — this only flips the default route selection and can be reverted with one env var once the driver accumulate matches the reference kernels on M5.
- Perf-sensitive paths should re-baseline M5 prefill numbers; the GLM Q2 A/B above showed no regression, but the DS MXFP4 workload that originally motivated the tensor route was not re-measured here.

### CODE REVIEW

**Comment by ivanfioravanti** [comment]
Closing until the write-up and change are reviewed internally. Will reopen (branch stays on the fork).
